src_feedback/elf_cache.rs

```diff
@@ -0,0 +1,304 @@
+    pub fn is_elf(&self) -> bool {
+	matches!(self, Self::Elf(_))
+    }
+}
+
+struct ElfCacheEntry {
+    // LRU links
+    prev: *mut ElfCacheEntry,
+    next: *mut ElfCacheEntry,
+
+    file_name: String,
+
+    dev: u64,
+    inode: u64,
+    size: i64,
+    mtime_sec: i64,
+    mtime_nsec: i64,
+    backend: ElfBackend,
+}
+
+impl ElfCacheEntry {
+    pub fn new(file_name: &str, file: File) -> Result<ElfCacheEntry, Error> {
+	let stat = fstat(file.as_raw_fd())?;
+	let parser = Rc::new(Elf64Parser::open_file(file)?);
+	let backend =
+	    if let Ok(dwarf) = DwarfResolver::from_parser_for_addresses(Rc::clone(&parser), &[]) {
+		ElfBackend::Dwarf(Rc::new(dwarf))
+	    } else {
+		ElfBackend::Elf(parser)
+	    };
+
+	Ok(ElfCacheEntry {
+	    prev: ptr::null_mut(),
+	    next: ptr::null_mut(),
+	    file_name: String::from(file_name),
+	    dev: stat.st_dev,
+	    inode: stat.st_ino,
+	    size: stat.st_size,
+	    mtime_sec: stat.st_mtime,
+	    mtime_nsec: stat.st_mtime_nsec,
+	    backend,
+	})
+    }
+
+    fn get_key(&self) -> String {
+	self.file_name.clone()
+    }
+
+    fn is_valid(&self, stat: &FileStat) -> bool {
+	stat.st_dev == self.dev &&
+	stat.st_ino == self.inode &&
+	stat.st_size == self.size &&
+	stat.st_mtime == self.mtime_sec &&
+	stat.st_mtime_nsec == self.mtime_nsec
+    }
+
+    fn get_backend(&self) -> ElfBackend {
+	self.backend.clone()
+    }
+}
+
+/// Maintain a LRU linked list of entires
+struct ElfCacheLru {
+    head: *mut ElfCacheEntry,
+    tail: *mut ElfCacheEntry,
+}
+
+impl ElfCacheLru {
+    /// # Safety
+    ///
+    /// Make all entries are valid.
+    unsafe fn touch(&mut self, ent: &ElfCacheEntry) {
+	self.remove(ent);
+	self.push_back(ent);
+    }
+
+    /// # Safety
+    ///
+    /// Make all entries are valid.
+    unsafe fn remove(&mut self, ent: &ElfCacheEntry) {
+	let ent_ptr = ent as *const ElfCacheEntry as *mut ElfCacheEntry;
+	let prev = (*ent_ptr).prev;
+	let next = (*ent_ptr).next;
+	if !prev.is_null() {
+	    (*prev).next = next;
+	} else {
+	    self.head = next;
+	}
+	if !next.is_null() {
+	    (*next).prev = prev;
+	} else {
+	    self.tail = prev;
+	}
+    }
+
+    /// # Safety
+    ///
+    /// Make all entries are valid.
+    unsafe fn push_back(&mut self, ent: &ElfCacheEntry) {
+	let ent_ptr = ent as *const ElfCacheEntry as *mut ElfCacheEntry;
+	if self.head.is_null() {
+	    (*ent_ptr).next = ptr::null_mut();
+	    (*ent_ptr).prev = ptr::null_mut();
+	    self.head = ent_ptr;
+	    self.tail = ent_ptr;
+	} else {
+	    (*ent_ptr).next = ptr::null_mut();
+	    (*self.tail).next = ent_ptr;
+	    (*ent_ptr).prev = self.tail;
+	    self.tail = ent_ptr;
+	}
+    }
+
+    /// # Safety
+    ///
+    /// Make all entries are valid.
+    unsafe fn pop_head(&mut self) -> *mut ElfCacheEntry {
+	let ent = self.head;
+	if !ent.is_null() {
+	    self.remove(&*ent);
+	}
+	ent
+    }
+}
+
+struct _ElfCache {
+    elfs: HashMap<ElfCacheEntryKey, Box<ElfCacheEntry>>,
+    lru: ElfCacheLru,
```

Comment:
Can you explain what the lru is used for within the ElfCache?

Reply:
ElfCacheLru  maintain a max capacity of the cache. Please check _ElfCach::ensure_size() & ElfCacheLru::touch().
Once the number of entries reaches a threshold, it purges least recently used entries to keep the size below the threshold.